### ISSUE TITLE
[JENKINS-17411] Pushing a fast-forward merge does not trigger a build

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/git-plugin#2267]
<p>I have a single job configured to build all branches of a git repo and poll for changes, and I have been experiencing a bug with build triggering.</p>

<h3><a name="Stepstoreproduce"></a>Steps to reproduce</h3>

<ol>
	<li>Create a new repo locally and run these in the repo:

<pre>
touch somefile.txt
git commit -m <span class="code-quote">"Add a file"</span> somefile.txt
</pre></li>
	<li>Create a new repo on github</li>
	<li>Create a Jenkins job with this configuration:
	<ul>
		<li>Set 'Repository URL' to the new github repo</li>
		<li>Set 'Branches to build' to '**'</li>
		<li>Check 'Prune remote branches before build', 'Skip internal tag', 'Clean after checkout', 'Recursively update submodules', 'Wipe out workspace before build'</li>
		<li>Check 'Poll SCM'</li>
		<li>Set 'Schedule' to '* * * * *'</li>
		<li>Add the following shell script as a build step:

<pre>
echo $GIT_BRANCH</pre></li>
	</ul>
	</li>
	<li>Push the master branch of your local repo to github</li>
	<li>Wait for Jenkins build (build #1)</li>
	<li>Run these in your local repo:

<pre>
git checkout -b feature master
echo stuff > somefile.txt
git commit -m <span class="code-quote">"Some changes"</span> somefile.txt
git push origin feature
</pre></li>
	<li>Wait for Jenkins build (build #2)</li>
	<li>Run these in your local repo:

<pre>
git checkout master
git merge feature
git push origin master
</pre></li>
</ol>

<h3><a name="Whathappens"></a>What happens</h3>

<p>After step 5, build #1 prints 'origin/master'.  After step 7, build #2 prints 'origin/feature'.  After step 8, nothing happens in Jenkins.</p>

<h3><a name="Whatshouldhappen"></a>What should happen</h3>

<p>Builds #1 and #2 are fine.  There should be a build #3 after step 8 which prints 'origin/master'.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nriffe">nriffe</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-17411" target="_blank">Pushing a fast-forward merge does not trigger a build</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ndeloof.png" title="ndeloof's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ndeloof">ndeloof</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: git-plugin
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2013-10-22T14:56:18+00:00
<li><b>votes</b>: 4
<li><b>watchers</b>: 6
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>I have a single job configured to build all branches of a git repo and poll for changes, and I have been experiencing a bug with build triggering.</p>

<h3><a name="Stepstoreproduce"></a>Steps to reproduce</h3>

<ol>
	<li>Create a new repo locally and run these in the repo:
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">touch somefile.txt
git commit -m <span class="code-quote">"Add a file"</span> somefile.txt
</pre>
</div></div></li>
	<li>Create a new repo on github</li>
	<li>Create a Jenkins job with this configuration:
	<ul>
		<li>Set 'Repository URL' to the new github repo</li>
		<li>Set 'Branches to build' to '**'</li>
		<li>Check 'Prune remote branches before build', 'Skip internal tag', 'Clean after checkout', 'Recursively update submodules', 'Wipe out workspace before build'</li>
		<li>Check 'Poll SCM'</li>
		<li>Set 'Schedule' to '* * * * *'</li>
		<li>Add the following shell script as a build step:
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">echo $GIT_BRANCH</pre>
</div></div></li>
	</ul>
	</li>
	<li>Push the master branch of your local repo to github</li>
	<li>Wait for Jenkins build (build #1)</li>
	<li>Run these in your local repo:
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">git checkout -b feature master
echo stuff &gt; somefile.txt
git commit -m <span class="code-quote">"Some changes"</span> somefile.txt
git push origin feature
</pre>
</div></div></li>
	<li>Wait for Jenkins build (build #2)</li>
	<li>Run these in your local repo:
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">git checkout master
git merge feature
git push origin master
</pre>
</div></div></li>
</ol>

<h3><a name="Whathappens"></a>What happens</h3>

<p>After step 5, build #1 prints 'origin/master'.  After step 7, build #2 prints 'origin/feature'.  After step 8, nothing happens in Jenkins.</p>

<h3><a name="Whatshouldhappen"></a>What should happen</h3>

<p>Builds #1 and #2 are fine.  There should be a build #3 after step 8 which prints 'origin/master'.</p></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>dant3:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-17411?focusedId=180458&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-180458">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Same here. Our environment uses automated deployment based on branching. As we merge changes to master we expect Jenkins to build them in master providing new artefact independently of is this the same SHA as in other branch, that Jenkins builded earlier. There is no way to choose such strategy atm.</p>
</pre>
</details>

</ul>
</details>

<p>Same here. Our environment uses automated deployment based on branching. As we merge changes to master we expect Jenkins to build them in master providing new artefact independently of is this the same SHA as in other branch, that Jenkins builded earlier. There is no way to choose such strategy atm.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/ndeloof.png" title="ndeloof's avatar" /> ndeloof:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-17411?focusedId=187921&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-187921">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>There is no actual distinction between "feature" and the fast-forwarded merge on "master". This is what merge-commit are designed for, if your workflow explicitly expect to track such merges</p>
</pre>
</details>

</ul>
</details>

<p>There is no actual distinction between "feature" and the fast-forwarded merge on "master". This is what merge-commit are designed for, if your workflow explicitly expect to track such merges</p>
